File: algorithm/programmers/LV2/swift/소수 찾기.swift
//
//  소수 찾기.swift
//  algorithm
//
//  Created by Kihun SONG on 2022/06/03.
//

import Foundation

// 순열로 숫자 조합 후 중복 제거를 위해 세트로 저장
func permute(_ array: [String], _ targetNumber: Int) -> Set<Int> {
	var result = Set<Int>()
	var visited = [Bool](repeating: false, count: array.count)

	func permutation(_ nowPermute: [String]) {
		if nowPermute.count == targetNumber {
			let temp = Int(nowPermute.joined())!
			result.insert(temp)
			return
		}
		for i in 0..<array.count {
			if visited[i] == true {
				continue
			} else {
				visited[i] = true
				permutation(nowPermute  + [array[i]])
				visited[i] = false
			}
		}
	}

	permutation([])

	return result
}

// 소수 판별
func isPrimeNumber(_ a: Int) -> Bool {
	if a <= 1 {
		return false
	}
	if a == 2 {
		return true
	}
	var i = 2
	while i <= (a / i) {
		if a % i == 0 {
			return false
		}
		i += 1
	}
	return true
}

func solution(_ numbers:String) -> Int {
	var count = 0
	let array = Array(numbers).map { String($0) }
	let arrayInt = array.map { Int($0)! }
	var setA = Set(arrayInt)

	// i 자리 수의 숫자세트를 만들고 합집합
	for i in 2...array.count {
		let tempSet = permute(array, i)
		setA = tempSet.union(setA)
	}

	for i in setA {
		if isPrimeNumber(i) {
			count += 1
		}
	}

	return count
}

File: algorithm/programmers/LV2/python/소수찾기.py
def solution(numbers):
    answer = 0
    result = set()

    number_list = list(numbers)

    def permutation(target, arr):
        result = []
        visited = [False for _ in range(len(arr))]

        def nowPermute(nowArr):
            if len(nowArr) == target:
                if nowArr not in result:
                    result.append(nowArr)
                return
            for i in range(len(arr)):
                if not visited[i]:
                    visited[i] = True
                    nowPermute(nowArr + [arr[i]])
                    visited[i] = False

        nowPermute([])
        return result

    def is_prime_number(number):
        if number == 0 or number == 1:
            return False

        for i in range(2, int(number**0.5)+1):
            if number % i == 0:
                return False
        return True

    for i in range(1, len(number_list)+1):
        permute_list = permutation(i, number_list)
        for element in permute_list:
            result.add(int("".join(element)))

    for number in list(result):
        if is_prime_number(number):
            answer += 1

    return answer
